ndcube/ndcube.py

```diff
@@ -607,11 +610,108 @@ def _fill_in_crop_nones(self, lower_corner, upper_corner, wcs, crop_by_values):
 
         return lower_corner, upper_corner
 
-    def _bounding_box_to_points(self, lower_corner_values, upper_corner_values, wcs):
+    def _bounding_box_to_points(self, lower_corner_values, upper_corner_values,
+                                axis_correlation_matrix):
         """
         Convert two corners of a bounding box to the points of all corners.
         """
-        return tuple(itertools.product(*zip(lower_corner_values, upper_corner_values)))
+        # Calculate which world axes share multiple pixel axes.
+        world_n_dep_dim = axis_correlation_matrix.sum(axis=1)
+        # If all world coordinates are independent, bounding box is simple.
+        max_dep_dim = world_n_dep_dim.max()
+        if max_dep_dim < 2:
+            return (tuple(lower_corner_values), tuple(upper_corner_values))
+        # If the number of world and pixel dims is the same, bounding box is also simple.
+        world_n_dim, pixel_n_dim = axis_correlation_matrix.shape
+        if world_n_dim == pixel_n_dim:
+            return tuple(itertools.product(*zip(lower_corner_values, upper_corner_values)))
+
+        # Otherwise we need calculate the corners more carefully.
+        # This must be done based on correlation matrix as not all
+        # world axes are independent.
+        # Start by generating array of world corner values assuming
+        # all coordinates are 1D. Strip units and add them back on later.
+        corner_units = [None] * len(lower_corner_values)
+        for i, (lcv, ucv) in enumerate(zip(lower_corner_values, upper_corner_values)):
+            if not isinstance(ucv, type(lcv)):
+                raise TypeError("Corresponding entries in lower and upper corner values "
+                                "must be of same type.")
+            if isinstance(lcv, u.Quantity):
+                corner_units[i] = lcv.unit
+                lower_corner_values[i] = lower_corner_values[i].value
+                upper_corner_values[i] = ucv.to_value(corner_units[i])
+        lower_corner_values = np.asarray(lower_corner_values)
+        upper_corner_values = np.asarray(upper_corner_values)
+        corners = np.stack([lower_corner_values, upper_corner_values])
+
+        # Next, calculate the sets of pixel axes upon which each world axis depends.
```

Comment:
```suggestion

        # Otherwise we need calculate the corners more carefully.
        # This must be done based on correlation matrix as not all
        # world axes are independent.
        # Start by generating array of world corner values assuming
        # all coordinates are 1D. Strip units and add them back on later.
        corner_units = [None] * len(lower_corner_values)
        for i, (lcv, ucv) in enumerate(zip(lower_corner_values, upper_corner_values)):
            if not isinstance(ucv, type(lcv)):
                raise TypeError("Corresponding entries in lower and upper corner values "
                                "must be of same type.")
            if isinstance(lcv, u.Quantity):
                corner_units[i] = lcv.unit
                lower_corner_values[i] = lower_corner_values[i].value
                upper_corner_values[i] = ucv.to_value(corner_units[i])
        lower_corner_values = np.asarray(lower_corner_values)
        upper_corner_values = np.asarray(upper_corner_values)
        corners = np.stack([lower_corner_values, upper_corner_values])

        # Next, calculate the sets of pixel axes upon which each world axis depends.
        world_n_dim, pixel_n_dim = axis_correlation_matrix.shape
```